### PR TITLE
Implement Couch Search and Request Batching

### PR DESCRIPTION
**Pull request by akhenry**
Fixes https://github.com/nasa/openmct/issues/3218
Fixes https://github.com/nasa/openmct/issues/3757

This PR implements some changes aimed at improving overall application performance in deployments that use Couch DB for persistence.

1. Object search now uses Couch DB, meaning those objects are no longer indexed by our in-memory indexer. This removes the need to do an exhaustive tree traversal on load to index all of the objects persisted in Couch.
2. `get`s for objects persisted in Couch DB are now debounced and batched, reducing the number of requests needed to retrieve objects from Couch DB.
3. The legacy Couch DB provider has been removed.
4. Made a small change in how search results are displayed. Was previously blocking on a large number of asynchronous calls before it would show any results.
5. Includes some small changes to the new and legacy object APIs to avoid redundant persistence calls when the modified and persisted timestamps are mutated.

### CODE REVIEW

File: src/plugins/persistence/couch/CouchObjectProvider.js [hunk 1]
```diff
@@ -131,10 +144,118 @@ export default class CouchObjectProvider {
...
+
+    /**
+     * @private
+     */
+    removeDuplicates(array) {
+        return Array.from(new Set(array));
+    }
+
+    search(query, abortSignal) {
+        const filter = {
```
Comment (shefalijoshi) [inline src/plugins/persistence/couch/CouchObjectProvider.js:245]: This search could return objects that have been removed from the tree previously (orphans?) resulting in invalid items in the search results.
We need to handle this.

File: src/ui/layout/mct-tree.vue [hunk 2]
```diff
@@ -710,25 +710,25 @@ export default {
                 }
             });
         },
-        async aggregateSearchResults(results, abortSignal) {
+        aggregateSearchResults(results, abortSignal) {
```
Comment (shefalijoshi) [inline src/ui/layout/mct-tree.vue:713]: Any specific reason to get rid of the async/await here?

**Review by shefalijoshi — CHANGES_REQUESTED** [review]

File: src/adapter/services/LegacyObjectAPIInterceptor.js [hunk 3]
```diff
@@ -52,7 +52,7 @@ define([
 
             oldStyleObject.getCapability('mutation').mutate(function () {
                 return utils.toOldFormat(newStyleObject);
-            });
+            }, newStyleObject.modified);
```
Comment (shefalijoshi) [inline src/adapter/services/LegacyObjectAPIInterceptor.js:55]: Ah cool! I had made a change for this in MutationCapablity.js but this is probably a better spot.

File: src/plugins/persistence/couch/CouchObjectProvider.js (hunk 1 above)
Comment (akhenry) [reply src/plugins/persistence/couch/CouchObjectProvider.js:245]: That's a great point. We have the same issue with our ElasticSearch plugin (which also supports server-side search).

I can't think of a reliable and performant way of handling this without an exhaustive tree traversal. I think the answer is to stop producing orphans at all by redesigning deletion, but even that is fraught with challenges. 

The way our object links work right now there is no easy way that I can think of to know whether an object is an orphan without an exhaustive tree traversal to see whether any other objects reference it.

Can we defer solving this problem to a followup issue?

File: src/ui/layout/mct-tree.vue (hunk 2 above)
Comment (akhenry) [reply src/ui/layout/mct-tree.vue:713]: The function doesn't return a `Promise` and it no longer contains any `await` expressions (they were blocking search results from rendering). `Await`-ing on this function would be meaningless.

File: src/plugins/persistence/couch/CouchObjectProvider.js (hunk 1 above)
Comment (shefalijoshi) [reply src/plugins/persistence/couch/CouchObjectProvider.js:245]: Yes. I've created this issue #3823

File: src/ui/layout/mct-tree.vue (hunk 2 above)
Comment (shefalijoshi) [reply src/ui/layout/mct-tree.vue:713]: Hehe... let me rephrase - You replaced the async/await with a .then() approach. I was curious about why?

Comment (akhenry) [reply src/ui/layout/mct-tree.vue:713]: Oh! Right, sorry. `getOriginalPath` triggers a bunch of requests to the server. Using `await` inside this array meant that each call to `getOriginalPath` could not even start until the previous one had resolved. This was blocking search results unnecessarily, and preventing batching.

Comment (shefalijoshi) [reply src/ui/layout/mct-tree.vue:713]: Ok cool thanks! :)

**Review by shefalijoshi — APPROVED** [review]

**Comment by shefalijoshi** [comment]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue) Y
